### ISSUE TITLE
unexpected error

### ISSUE THREAD

**Issue by serbring** [TypHo22/ConcaveHull#1]
Hi, 

what a great the function that performs way better than the built-in one. Have you tried to repeatedly run the function through a for-loop? On my code, it seems to be unstable for some reasons. Have you experienced any issue like that? 

All the best,

Pietro

**Comment by TypHo22** [comment]
Hey Pietro,

glad that you like it. No I have not tried it out. Can you give me an example? 

Best Regards, 

Andreas

**Comment by serbring** [comment]
Hi Andreas, 

thanks for your quick reply. In preparing a little code to show you, I have found some random behaviour of the function but limited in my script. I have not understood yet the reason but it is in my code. However, I have found that in certain conditions the function is very slow. Here you can find my code and the dataset I struggle with. Why there is this problem?
Thank you for your support. 

```
clear all
load("data4concavehull.mat")
[boundary] = concaveHull([x4convhullRes' y4convhullRes'] ,floor(length(x4convhullRes)./5));

```


[data4concavehull.zip](https://github.com/TypHo22/ConcaveHull/files/11952396/data4concavehull.zip)

**Comment by TypHo22** [comment]
I will investigate the issue the next days, thanks for sharing with me.

**Comment by TypHo22** [comment]
Hi Pietro @serbring ,

the algorithm is fine but runs into issues because its k-value is to small and/or your dataset consists of more  than a single group of points. 

The abortion criteria for the algorithm is that it finds its very first vertex again (The one with the lowest y-value). 
The first vertex is not found anymore because your k is to small and the first vertex is located in the first cluster which is not in reach anymore for suitable candidates (all candidates are from second cluster and filling up the amount of k-neighbours). 

For your particular case there are two things you can do: 

- choose a higher k value.
- clean by first cluster because those points seem to be more of outliers and create their own cluster


![problem](https://github.com/TypHo22/ConcaveHull/assets/42981587/20aae205-59f3-4c38-9f11-c651f71ea76a)



Running it with the following k works for me: 

`floor(length(x4convhullRes)./3)`


because you are talking of performance of the algorithm. 
Did you removed the lines for visualization? 

Every line which has a % for visualisation only
is not needed and slows the algorithm down (was for a showcase). 


does this answer your question?

**Comment by serbring** [comment]
Hi Andreas,

I had a look, I did a few testing (this lead a little delay in my answer, I am sorry for this) and the function is susceptible to the presence of different clusters in the point cloud. I will try to figure out how to isolate clusters but it is not easy in any case.



Best regards,



Pietro

Inviato da Posta<https://go.microsoft.com/fwlink/?LinkId=550986> per Windows

Da: ***@***.***>
Inviato: lunedì 10 luglio 2023 14:35
A: ***@***.***>
Cc: ***@***.***>; ***@***.***>
Oggetto: Re: [TypHo22/ConcaveHull] unexpected error (Issue #1)


Closed #1<https://github.com/TypHo22/ConcaveHull/issues/1> as not planned.
